server/src/main/java/org/apache/druid/catalog/model/table/HttpInputSourceDefn.java

```diff
@@ -0,0 +1,304 @@
+      // will pass.
+      try {
+        sourceMap.put(
+            URIS_FIELD,
+            Collections.singletonList(new URI("https://bogus.com/file"))
+        );
+      }
+      catch (Exception e) {
+        throw new ISE(e, "URI parse failed");
+      }
+    }
+    super.validate(table);
+  }
+
+  private Matcher templateMatcher(String uriTemplate)
+  {
+    Pattern p = Pattern.compile("\\{}");
+    Matcher m = p.matcher(uriTemplate);
+    if (!m.find()) {
+      throw new IAE(
+          "Value [%s] for property %s must include a '{}' placeholder",
+          uriTemplate,
+          URI_TEMPLATE_PROPERTY
+      );
+    }
+    return m;
+  }
+
+  @Override
+  protected List<ParameterDefn> adHocTableFnParameters()
+  {
+    return CatalogUtils.concatLists(URI_PARAMS, USER_PWD_PARAMS);
+  }
+
+  @Override
+  protected void convertArgsToSourceMap(Map<String, Object> jsonMap, Map<String, Object> args)
+  {
+    jsonMap.put(InputSource.TYPE_PROPERTY, HttpInputSource.TYPE_KEY);
+    convertUriArg(jsonMap, args);
+    convertUserPasswordArgs(jsonMap, args);
+  }
+
+  @Override
+  public TableFunction partialTableFn(ResolvedExternalTable table)
+  {
+    List<ParameterDefn> params = Collections.emptyList();
+
+    // Does the table define URIs?
+    Map<String, Object> sourceMap = table.inputSourceMap;
+    if (!sourceMap.containsKey(URIS_FIELD)) {
+      params = CatalogUtils.concatLists(params, URI_PARAMS);
+    }
+
+    // Does the table define a user or password?
+    if (!sourceMap.containsKey(USERNAME_FIELD) && !sourceMap.containsKey(PASSWORD_FIELD)) {
+      params = CatalogUtils.concatLists(params, USER_PWD_PARAMS);
+    }
+
+    // Does the table define a format?
+    if (table.inputFormatMap == null) {
+      params = addFormatParameters(params);
+    }
+    return new PartialTableFunction(table, params);
+  }
+
+  @Override
+  protected ExternalTableSpec convertCompletedTable(
+      final ResolvedExternalTable table,
+      final Map<String, Object> args,
+      final List<ColumnSpec> columns
+  )
+  {
+    // Get URIs from table if defined, else from arguments.
+    final Map<String, Object> sourceMap = new HashMap<>(table.inputSourceMap);
+    final String uriTemplate = table.resolvedTable().stringProperty(URI_TEMPLATE_PROPERTY);
+    if (uriTemplate != null) {
+      convertUriTemplateArgs(sourceMap, uriTemplate, args);
+    } else if (!sourceMap.containsKey(URIS_FIELD)) {
+      convertUriArg(sourceMap, args);
+    }
+
+    // Get user and password from the table if defined, else from arguments.
+    if (!sourceMap.containsKey(USERNAME_FIELD) && !sourceMap.containsKey(PASSWORD_FIELD)) {
+      convertUserPasswordArgs(sourceMap, args);
+    }
+    return convertPartialFormattedTable(table, args, columns, sourceMap);
+  }
+
+  private void convertUriTemplateArgs(Map<String, Object> jsonMap, String uriTemplate, Map<String, Object> args)
+  {
+    List<String> uriStrings = CatalogUtils.getStringArray(args, URIS_PARAMETER);
```

Comment:
Have we verified before this that the `URIS_PARAMETER` is present in the args, or can this lead to NPE somehow?